### PR TITLE
Reduce visual prominence of primary actions in table data views, and consolidate primary + secondary actions in ellipsis menu

### PR DESCRIPTION
**Pull request by jameskoster**
## What?
1. Reduce the permanent visual prominence of primary actions in table data views.
2. Consolidate primary + secondary actions into the ellipsis menu

## Why?
Permanently displaying the primary actions is quite noisy and adds a lot of repetition to the UI. Revealing them while focussing a specific row can better manage prominence. 

The ellipsis button remains permanently visible so that users can intuit that actions exist at a glance. 

Consolidating all actions into a single menu is helpful for orientation, and means we can use a single reference point for the right-click interaction (https://github.com/WordPress/gutenberg/issues/59608).

## How?
Primary actions are revealed using the same mechanics as the checkbox. That is; when a `tr` has `is-hovered` class, or has `focus-within` the actions appear. Secondary actions (via ellipsis menu) are permanently visible.

## Testing Instructions
* Open a data view such as Pages or Templates
* Select Table layout
* Observe that primary actions appear only when hovering a row, or when focus is within a row.

https://github.com/WordPress/gutenberg/assets/846565/312d8789-e8e9-403b-877e-ef5aaccfa469

**Note**: I'd say this is quite closely related to https://github.com/WordPress/gutenberg/pull/59551. The placement of the Trash action in the menu seems a little odd. This would make more sense:

– Edit
– View
– Revisions
– Trash

Ideally these PR's are reviewed in tandem.

### CODE REVIEW

**Comment by carolinan** [comment]
How can we make sure it is still discoverable?
Is there a "welcome guide" on this updated interface?

**Comment by carolinan** [comment]
Has there been any feedback from users about this interface being too cluttered?
I'm worried the change will make it more difficult to use.
For decisions like these it would be great to have some real test results...


How does this work for users of screen magnification software?
Without having any data to back this up, I believe that if you are hovering over anything in this interface, you are **most likely** to hover over the Title, as it is the most important and most prominent item.
In the wp_admin interface, the objects that show on hover are very close to the title.
Here the objects that show on hover are very far from what your are likely focusing your mouse and your sight on.

**Comment by afercia** [comment]
I'm not sure this proposed change helps usability and accessibility to get any better. Hiding content isn't the best way to make an user interface more usable. To me, the proposed change seems to aim to improve only the visuals, based on some subjective judgment, and doesn't help functionality. It seems to me that it prioritizes form over function, which is less than ideal IMHO.

I'd think we should go the opposite direction instead, at the point I wanted to create an issue to propose to make the checkboxes always visible instead of showing them only on hover or focus. I'm going to create that issue soon.

Besides that, I _have_ to say that I'm uncomfortable with the process followed with this PR. I'm not sure a pull request is the right place to discuss process but I'd like to clarify why I'm uncomfortable with this. I'll try to discuss this point in more depth in a more appropriate place and time. 

1. This PR is not preceded by an issue. The contributing guidelines of the Gutenberg project (and I'd say also the process followed in Core since ages) [state](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/repository-management.md#pull-requests) that: _Non-trivial pull requests should be preceded by a related issue that defines the problem to solve and allows for discussion of the most appropriate solution before actually writing code._ I think this is an essential point in any collaborative open source project. It's not the only PR without associated issue, there have been many such cases. Keeping ignoring this point doesn't encourage contribution and, honestly, is highly demotivating.
2. This PR changes an important aspect of the user interface but no accessibility label was set on it. Thankfully, @carolinan added the 'Needs accessibility feedback' label a few hours ago. However, making a relevant change to the UI without asking for _broad_ accessibility feedback isn't the best way to make sure the proposed change is inclusive. 
3. I see a few people have been pinged for a review of this PR including andrewhayward. Andrew is an excellent, trusted, and well respected accessibility specialist. However, I would like to remind everyone that contribution to WordPress is organized into teams. One of these teams is the Accessibility team. Not pinging the Accessibility team is, honestly, dismissive of the team's role and I perceive it as unfair. I'm assuming good intent but _the perception_ of the process on this PR isn't great.
4. This PR is marked as 'Enhancement'. It's also marked with the 'Backport to WP Beta/RC' label. I'm not sure any enhancement should be merged during Beta. Not to mention there's no agreement this is an actual enhancement, but my point is more about the process.

It's not my intent to blame the specific person who submitted this PR. As I mentioned above, there have been _several_ cases of pull requests that didn't follow the established process. I'm reporting a general issue that I think needs to be discussed looking at the broader picture. I apologize in advance if anyone may feel personally targeted by my feedback but that's not my intent.

**Comment by richtabor** [comment]
I'm curious, how does seeing a thumbnail with a label of "Thumbnail" more or less functional than one without a label? 

If it helps visually and does not affect otherwise, it is an improvement.

**Review by andrewhayward — COMMENTED** [review]
> Has there been any feedback from users about this interface being too cluttered?
> I'm worried the change will make it more difficult to use.

Just to echo @carolinan's thoughts, I'm not entirely sure as the motivation behind this is actually needed, or that it really improves things. Happy to discuss, regardless.

Aside from the various specific conceptual points brought up though, there are a couple of technical issues that we would need to resolve before advancing with this, should we do so.

A screen-grab showing table rows being hovered to reveal action components, with tooltips then appearing when those components are themselves hovered. The action components disappear again when the cursor is placed over the tooltip.

The various action components have associated tooltips, which themselves appear on hover. However, when hovering over the tooltips, the components then disappear. It's not quite [the typical failure](https://www.w3.org/WAI/WCAG21/Techniques/failures/F95) of [SC 1.4.13 _(Content on Hover or Focus)_](https://www.w3.org/WAI/WCAG22/Understanding/content-on-hover-or-focus), but it's a failure nevertheless!

A screen-grab showing table rows receiving focus to reveal action components, one of which being a menu button. When the menu is opened, the components disappear.

Similarly, when the "more actions" menu is activated, the action components disappear again. The illustration above shows only a keyboard focus failure, but the same issue applies with pointer users too.

**Comment by andrewhayward** [comment]
Happened to be looking at Apple Podcasts, and noticed they only hide the primary actions, and reveal them on hover; the actions menu itself is always visible, and also contains the primary actions. Not quite the same paradigm, and not necessarily suggesting that just because someone else does it so should we, but food for thought, regardless.

A screen capture of Apple Podcasts, showing action items appearing on hover, before opening the always-visible action menu to expose a longer list of actions, including the separate primary actions.

**Comment by carolinan** [comment]
Hi
is this PR still planned for 6.5? Kind reminder that RC1 is tomorrow, March 4. Perhaps the backport label should be removed?

**Comment by jameskoster** [comment]
@andrewhayward that pattern seems worth a try. It surfaces primary actions only when focussed on a specific row to manage prominence, and (hopefully) informs users that actions exist regardless, in a more minimal way.

**Comment by andrewhayward** [comment]
Plays a bit weird at the moment because of where menus are rendered in the DOM, and how we're relying on native hover/focus.

Action buttons disappearing when dropdown menu opened

**Comment by jameskoster** [comment]
You mean how the action buttons disappear when you engage the ellipsis menu? I think that weirdness also applies to the hover state.

As you alluded to, ideally an `is-hovered` class would be toggled, and persist when interacting with the menu. Then we could ensure the action buttons and styles remain visible.

**Comment by jameskoster** [comment]
> ideally an is-hovered class would be toggled, and persist when interacting with the menu.

I added this. It works a bit better now – the hover styles and primary action buttons remain visible when you interact with the ellipsis menu:

Screenshot 2024-03-04 at 16 27 56

**Comment by jameskoster** [comment]
I'd say this is quite closely related to https://github.com/WordPress/gutenberg/pull/59551. The placement of the Trash action in the menu seems a little odd. This would make more sense:

– Edit
– View
– Revisions
– Trash

**Comment by t-hamano** [comment]
I think more discussion is needed as to what approach is best, but I would like to introduce two apps. We may be able to get some hints from these apps.

### Google Drive

- The ellipsis menu is always displayed.
- The left button is only visible while hovering over a row.

https://github.com/WordPress/gutenberg/assets/54422211/971950b9-b160-42ba-b39b-054202c11258

### Dropbox

The left button and ellipsis menu are only visible while hovering over a row.

https://github.com/WordPress/gutenberg/assets/54422211/dcec1207-aa96-497d-9429-1e58a40bd6f3

**Comment by jasmussen** [comment]
Since this PR is 6.6 bound, I think it would be a great way to unstick this conversation to actually land it, and have feedback trickle in from its usage in the plugin.

As is, this behavior is identical to what exists in Google Drive, and what has existed in WordPress for many years:

Screenshot 2024-03-06 at 09 18 52

**Comment by jameskoster** [comment]
Yes this seems to be a common pattern: 

* Ellipsis menu contains the full list of actions. 
* The buttons that appear on row focus/hover are the most commonly-used, extracted from the ellipsis menu so they're accessible in a single click.  

This is a little different to the original convention we installed, but given we want to surface actions on right-click down the road (https://github.com/WordPress/gutenberg/issues/59608), I think this moves us in a better direction. It's also similar to the existing wp-admin lists as pointed out above, with added scalability.

**Comment by t-hamano** [comment]
One concern I have is that I don't know what primary actions are available for that item unless I hover over that row.

For example, on a Pages page, both "View" and "Move to Trash" actions are always visible for any item. The actions available for each item are unified, so it may not be a problem if it is hidden by default.

On the other hand, for example on a template page, the available actions change depending on the item's status. Sometimes there is no action, sometimes there is a reset or trash button. In this way, when the icon changes depending on the item's status, it is a little hesitant to hide the icon with the default.

https://github.com/WordPress/gutenberg/assets/54422211/98d51b06-1950-461c-b6ee-99ba05bb132e

**Comment by jameskoster** [comment]
@t-hamano I'm inclined to say that's a separate issue which also exists (albeit to a slightly lesser extent) on trunk. It's arguably confusing that different templates have different action buttons, regardless of how they appear. 

Solutions to that problem are perhaps best discussed in detail elsewhere, but options include; unifying the reset/delete actions, or making those secondary so all template actions live in the ellipsis menu.

**Comment by jasmussen** [comment]
I also don't completely buy the argument that it's hard to know, both because this is an existing behavior in WordPress with the on-hover quick-actions, but also because every action you can take is consistently available in the ellipsis menu.

**Comment by andrewhayward** [comment]
There are a couple of things we need to ensure we're aligned on, but otherwise I think it's worth moving ahead with this; we're more likely to get feedback if it's directly available to users, and this feels like one of those ones that we could go back and forth on all day.

1. The "extras" menu should always be visible; only the "shortcut" actions are hidden. It should be clear that actions are available even without hover/focus, and a persistent menu goes a long way to achieving that.
2. The "shortcut" actions should be (additionally) included in the "extras" menu. There needs to be one consistent place to find actions, and being able to cognitively target that menu means the top-level actions become secondary in importance; useful to have, but not essential for operation.

> ...this behavior is identical to what exists in Google Drive, what has existed in WordPress for many years

I'm always wary of copying patterns just because, but rightly or wrongly it's not uncommon. As @jasmussen points out, it already exists in WordPress, and this would actually be an improvement; while definitions of "perfect" vary, I'm all for small steps in the right direction.

**Comment by jameskoster** [comment]
Those points seem reasonable to me 👍

**Comment by andrewhayward** [comment]
> Those points seem reasonable to me 👍

If we're in agreement, I'm happy to proceed and iterate where necessary.

> I'd think we should go the opposite direction instead, at the point I wanted to create an issue to propose to make the checkboxes always visible instead of showing them only on hover or focus.

@afercia I'm largely in agreement with #59175, and it's definitely a conversation we should continue there. As this is a slightly different scenario, would love your thoughts on [the alignment points above](https://github.com/WordPress/gutenberg/pull/59128#issuecomment-1983576172) (cc @carolinan).

**Review by andrewhayward — APPROVED** [review]

**Comment by jameskoster** [comment]
Just leaving a note to say we should follow-up to evaluate action order, and which should have single-click access. The menu order in this PR is a little odd (Trash is given more prominence than Edit). I'll pick that up in https://github.com/WordPress/gutenberg/pull/59551.

**Comment by afercia** [comment]
@andrewhayward late to the party, but if you ask me, I'd make _everything_ always visible by default. If a 'ligher' UI is desirable, that should be _optional_. It's not uncommon in large applications that need to account for different needs to have layout options like 'detailed', 'compact' and so on. However, the default should be usable and accessible for everyone.
It's a broader topic, i know.

> what has existed in WordPress for many years

That's not entirely accurate. In core, the 'view actions' in the List Tables are revealed on hover / focus for Posts, Pages, Users, etc. but for Plugins they're always visible and that's been the case since forever, I think. Mentioning it just to better clarify some statements in this conversation aren't entirely accurate.

File: packages/dataviews/src/item-actions.js [hunk 1]
```diff
@@ -168,10 +170,16 @@ export default function ItemActions( { item, actions, isCompact } ) {
 						icon={ moreVertical }
 						label={ __( 'Actions' ) }
 						disabled={ ! secondaryActions.length }
+						className="dataviews-view-table__all-actions-button"
```
Comment (ntsekouras) [inline packages/dataviews/src/item-actions.js:173]: We shouldn't use classnames for specific views in the generic component. I have opened a follow up [PR](https://github.com/WordPress/gutenberg/pull/59748) for this.

**Pull request merged**